### ISSUE TITLE
Can't scale the preview window with flash

### ISSUE THREAD

**Issue by ErikEklund** [jhuckaby/webcamjs#146]
I would like to get a stream of 1280x720, crop it to 720x720, and display the resulting video stream in a smaller size - say 400x400 - while the final output image is still 720x720. An option to accomplish this would be great.

It is possible to scale the container in css with transform: scale(), but it seems flash don't take scaling into account - you can't click the accept/deny buttons. It would be better if you could pick a scale (or width/height) in the options, so that the resulting div is properly sized.

**Comment by jhuckaby** [comment]
See the below excerpt from the docs -- does this answer your question?  Or are you talking about something else?

## Cropping The Image

WebcamJS can also crop the final image for you, to any dimensions you like.  This is useful for when you want a square image (perhaps for a website profile pic), but you want to capture the image from the user's webcam at 4:3 ratio to be fully compatible (some cameras require 4:3 and cannot capture square images).  To do this, include `crop_width` and `crop_height` params, specifying the area to crop out of the center of the final image:

```javascript
	Webcam.set({
		width: 320,
		height: 240,
		crop_width: 240,
		crop_height: 240
	});
	
	// Attach camera here
```

This would crop a 240x240 square out of the center of the 320x240 webcam image.  The crop is also reflected in the live preview area.  In this case the live preview would also be cropped to 240x240, so the user can see exactly what portion of the image will be captured.

[See a live demo of this feature here](https://pixlcore.com/demos/webcamjs/demos/crop.html)

**Comment by ErikEklund** [comment]
Not really. I want to do crop _and_ scale at the same time. Say something like this:

    Webcam.set({
        // Size of stream from camera
        width: 1280,
        height: 720,

        // Size of crop
        crop_width: 720,
        crop_height: 720,

        // Display size of the DOM element displaying the content
        display_width: 300,
        display_height: 300
    });

I don't know if I'm misreading the docs, but I've tried most combinations of width/height, crop_width/crop_height and displ_width/disp_height I can think of. Right now it seems the resolution of the video stream is tightly integrated with the size of the output. Ie, currently I can't seem to take a 1280x720 stream, crop it to 720x720 and display it at a lower resolution than 720x720.

**Comment by jhuckaby** [comment]
I don't believe `display_width` and `display_height` are valid parameter names.

This demo does both a crop and a scale:

https://pixlcore.com/demos/webcamjs/demos/crop-large.html

I think the key is in setting up the parameters.  You need to make sure that the aspect ratio of the `width` and `height` params matches that of the `dest_width` and `dest_height` ratio.  Then you can set the `crop_width` and `crop_height` to whatever ratio you want.  Like this:

```js
Webcam.set({
	// live preview size
	width: 320,
	height: 240,
			
	// device capture size
	dest_width: 640,
	dest_height: 480,
			
	// final cropped size
	crop_width: 480,
	crop_height: 480,
			
	// format and quality
	image_format: 'jpeg',
	jpeg_quality: 90
});
```

Let me know if this still doesn't do the trick.  I will admit, I've never tested this at HD resolutions with the Flash fallback.

**Comment by ErikEklund** [comment]
Hm. Yes and no. I think I see the issue now. I use the following config:

    Webcam.set({
        width: 400,
        height: 400,
        dest_width: 720,
        dest_height: 720,
        crop_width: 720,
        crop_height: 720,
        constraints: {
            mandatory: {
                minWidth: 640,
                minHeight: 480
            }
        },
        force_flash: false
    });

I get a 400x400 element with a 16:9 uncropped video inside of it, and bars at the top and bottom. This was what I was trying to fix in my previous attempts. Though now I noticed that using force_flash: true gives the desired result, ie a 400x400 element where the video fills the entire area. It would be good if they behaved the same.

**Comment by jhuckaby** [comment]
Yeah, your aspect ratios don't match.  I believe it gets really confused if the `width / height` doesn't match the `dest_width / dest_height`.  Sorry about the HTML5 vs. Flash not behaving the same.  It's very difficult to make it consistent in all cases (and with all cameras and drivers).  I'll leave this open as a bug, and try to see if I can fix it when I have some time.

**Comment by ErikEklund** [comment]
No, I added the constrains only to test this actually. Even without constraints, the HTML version renders with borders. There seems to be some slight issue with scaling/cropping, at least on the HTML side.

You might want to relabel this bug if you're leaving it open - I started with making the HTML look good, and then try to fix the flash, which also have some headaches. It's not a very descriptive title, now that we nailed down the issue :).

I can use it with force_flash on for now, so all good. Thanks a lot for fast feedback and for an awesome piece of code :).
